tool/tsh/common/tsh.go

```diff
@@ -3225,17 +3248,58 @@ func accessRequestForSSH(ctx context.Context, _ *CLIConf, tc *client.TeleportCli
 
 	// At this point we have exactly 1 node.
 	node := rsp.Servers[0]
+	var req types.AccessRequest
 	requestResourceIDs := []types.ResourceID{{
 		ClusterName: tc.SiteName,
 		Kind:        types.KindNode,
 		Name:        node.GetName(),
 	}}
+	switch cf.RequestMode {
+	case accessRequestModeRole:
+		rootClient, err := clt.ConnectToRootCluster(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		resp, err := rootClient.GetAccessCapabilities(ctx, types.AccessCapabilitiesRequest{
+			RequestableRoles:                 true,
+			ResourceIDs:                      requestResourceIDs,
+			Login:                            tc.HostLogin,
+			FilterRequestableRolesByResource: true,
+		})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
 
-	// Roles to request will be automatically determined on the backend.
-	req, err := services.NewAccessRequestWithResources(tc.Username, nil /* roles */, requestResourceIDs)
-	if err != nil {
-		return nil, trace.Wrap(err)
+		var selectedRole string
+		switch len(resp.RequestableRoles) {
+		case 0:
+			return nil, trace.AccessDenied("no roles to request that would grant access")
+		case 1:
+			selectedRole = resp.RequestableRoles[0]
+		default:
+			selectedRole, err = prompt.PickOne(
+				ctx, os.Stdout, prompt.NewContextReader(os.Stdin),
+				"Choose role to request",
+				resp.RequestableRoles)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
+
+		req, err = services.NewAccessRequestWithResources(tc.Username, []string{selectedRole}, nil /* requestResources */)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	case accessRequestModeResource:
+		// Roles to request will be automatically determined on the backend.
+		req, err = services.NewAccessRequestWithResources(tc.Username, nil /* roles */, requestResourceIDs)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	default:
+		return nil, trace.BadParameter("unexpected request mode %q", cf.RequestMode)
 	}
+
 	req.SetLoginHint(tc.HostLogin)
```

Comment:
I don't think this and the DryRun request below are necessary if we're doing a role request, `GetAccessCapabilities` serves the same purpose. Can we skip it to save a round-trip? Let's only do this in the `accessRequestModeResource` case, and consider making each case its own function